Fix autosave while editing a post using the Text Mode editor

noisysocks opened the pull request:
## Description
Fixes #2978. 

Text that was typed into the Text Mode editor would be wiped away if the post was autosaved. Here's what was happening:

1. A user switches to Text Mode and makes changes. The modified content is stored in `editor.present.edits.content`. Note that `editor.present.blocksByUid` is not modified until the text area is blurred.
2. After some time, the post is autosaved. When the successful save request finishes, `REQUEST_POST_UPDATE_SUCCESS` is dispatched.
3. `REQUEST_POST_UPDATE_SUCCESS` dispatches `RESET_POST`, which clears out the edits in `editor.present.edits`.
4. Now that `editor.present.edits.content` is undefined, `getEditedPostContent( state )` returns a serialisation of the blocks in `editor.present.blocksByUid`. But, since the editor was never blurred, this does not reflect the recent changes that were made! We therefore lose those changes.

I've fixed this by patching `PostTextEditor` so that it copies `value` into local state when focused. This means that the contents of the text field won't change if `value` changes because of a dispatch.

## How Has This Been Tested?
1. Write a new post in Gutenberg
2. Switch to Text Mode
3. Type in a bunch of stuff
4. Wait for autosave to kick in—your changes should remain there
6. Refresh the page—your changes should remain there
8. Switch back to Visual Mode—your changes should appear

noisysocks committed:
Prevent changes from being lost while in Text Mode

Store the value of the PostTextEditor text area in local state. This
prevents text from being lost if the post is autosaved during editing.

noisysocks commented:
This fixes the bug but I find it a bit weird that there are moments where `editor.present.blocksByUid` does not reflect what's currently stored in `currentPost.content`.

I'm considering changing `PostTextEditor` so that it calls `resetBlocks( parse( content ) )` when the text field is _changed_ instead of blurred. This way, the two subtrees of our editor state remain in sync. Debouncing the change event should keep the editor feeling responsive.

Thoughts?

youknowriad commented:
> I'm considering changing PostTextEditor so that it calls resetBlocks( parse( content ) )

Yes, this is a hard problem to solve with constant syncing. We tried to avoid this for performance reasons. Maybe it's fine I don't know, we have to try for long posts as well. cc @aduth

aduth commented:
Previously it would be an issue because parsing an incomplete text would probably alter the serialized output of the reset blocks. Using state to ensure that the textarea value isn't clobbered should remedy this as far as the user is concerned, but the parsed data is probably still often out of sync with what the user is typing.

aduth commented on editor/components/post-text-editor/index.js:
```diff
@@ -18,41 +18,45 @@ import { getEditedPostContent } from '../../store/selectors';
 
-	onChange( event ) {
-		this.props.onChange( event.target.value );
+	handleFocus() {
```
FWIW I find these sorts of function names aren't very informative and encourage developers to not clearly separate the responsibilities of functions. As an outsider looking at this function later, I'd have no idea what it's intending to do, at least not without reading through and interpreting the logic (and who has time for that?). If it were named `setStateValue` with a DocBlock explaining how we track a local copy of the value to avoid clobbering edits from global state, I'd feel much more informed.

There's a few places where I've tried to set more explicit patterns for some of these common behaviors. One in particular is handling `onKeyDown`, since we usually want distinct functionality to take place on individual key codes. While maybe a bit extreme (especially since it only handles as a single key code _at the moment_), `NavigableToolbar` is one such example:

https://github.com/WordPress/gutenberg/blob/4bac4fe5e4f918d1ba24ee12e8f00f6fcc89b49c/editor/components/navigable-toolbar/index.js#L32-L34

https://github.com/WordPress/gutenberg/blob/4bac4fe5e4f918d1ba24ee12e8f00f6fcc89b49c/editor/components/navigable-toolbar/index.js#L51-L73

https://github.com/WordPress/gutenberg/blob/4bac4fe5e4f918d1ba24ee12e8f00f6fcc89b49c/editor/components/navigable-toolbar/index.js#L83

(And yes, arguments can be made that `_.cond` is not a very obvious / readable utility, but better abstractions could exist. The general point stands all the same)

noisysocks replied:
👍 noted. I've moved away from 'handle-' method names in c4035be7a.

aduth commented on editor/components/post-text-editor/index.js:
```diff
@@ -18,41 +18,45 @@ import { getEditedPostContent } from '../../store/selectors';
-				initialValue: value,
-			} );
+	handleBlur() {
+		if ( this.state.isDirty ) {
```
Do we need to track `isDirty`? Can't we just check `if ( this.state.value !== this.props.value ) {` ?

noisysocks replied:
Since we change the global value by dispatching `EDIT_POST` in our `onChange` handler, 
`this.state.value !== this.props.value` will always evaluate to false.

noisysocks committed:
Avoid 'handle-' method names

Try to describe what the method does instead of merely what event it
handles.

noisysocks commented:
> Using state to ensure that the textarea value isn't clobbered should remedy this as far as the user is concerned, but the parsed data is probably still often out of sync with what the user is typing.

Yeah, makes sense. Let's stick with this approach for now, then.

noisysocks commented:
Are we good to go ahead and get this in?

youknowriad reviewed (approved):
What about external updates, in the current implementation they are ignored completely, I think an option could be:

 - If we receive new content and the editor is not dirty,  replace the state value if existant

 - If we receive new content and the editor is dirty, an option would be to compare the values and if different show a notice saying if you want to revert changes (new external updates) you can.

The second point can probably be done later, I see its value especially for plugins that could update the content independently from the UI. Granted we don't have this use case right now.

noisysocks committed:
Don't ignore *all* external updates in the post text editor

If the post text editor isn't dirty, update the editor with any external
updates.

noisysocks commented:
> If we receive new content and the editor is not dirty, replace the state value if existant

I've implemented this in af209c25f and tested it by:

1. Running this in the DevTools console:
  ```js
  setTimeout( () => { wp.data.dispatch( 'core/editor' ).editPost( { content: '<!-- wp:paragraph -->\n<p>hey there! How are you?</p>\n<!-- /wp:paragraph -->\n\n<!-- wp:paragraph -->\n<p>Pretty well, thank you</p>\n<!-- /wp:paragraph -->' } ) }, 2000 )
  ```
2. Quickly focusing the text area with my mouse. If I don't type anything, the contents will change after 2 seconds
